message_ix/macro.py

```diff
@@ -1,61 +1,471 @@
+        partmp = (bconst * demand_ref ** rho).groupby(level='node').sum()
+        # TODO: automatically get the units here!!
+        aconst = ((gdp0 / 1e3) ** rho - partmp) / (k0 / 1e3) ** (rho * kpvs)
+        # want the series to only have index of node
+        self.data['aconst'] = aconst.reset_index(level='year', drop=True)
+        return self.data['aconst']
+
 
 def init(s):
-    for args in MACRO_INIT['sets']:
-        s.init_set(*args)
-    for args in MACRO_INIT['pars']:
-        s.init_par(*args)
-    # for args in MACRO_INIT['vars']:
-    #     print(args)
-    #     s.init_var(*args)
-    # for args in MACRO_INIT['equs']:
-    #     s.init_equ(*args)
+    for key, values in MACRO_INIT['sets'].items():
+        s.init_set(key, values)
+    for key, values in MACRO_INIT['pars'].items():
+        try:
+            s.init_par(key, values['idx'])
+        except:  # noqa: ignore=E722
+            pass  # already exists in the model, known for 'historical_gdp'
+    for key, values in MACRO_INIT['vars'].items():
+        if not s.has_var(key):
+            try:
+                # TODO: this seems required because for some reason DEMAND (and
+                # perhaps others) seem to already be listed in the java code,
+                # but still needs to be initialized in the python code.
+                # However, you cannot init it with dimensions, only with the
+                # variable name.
+                s.init_var(key, values)
+            except:  # noqa: ignore=E722
+                s.init_var(key)
+    for key, values in MACRO_INIT['equs'].items():
+        s.init_equ(key, values)
+
+    # keep track of number of iterations
+    s.init_var('N_ITER', None)
+    s.init_var('MAX_ITER', None)
+
+
+def add_model_data(base, clone, data):
+    c = Calculate(base, data)
+    c.read_data()
+    c.derive_data()
+
+    # add temporal set structure
+    clone.add_set("type_year", "initializeyear_macro")
+    clone.add_set("cat_year", ["initializeyear_macro", c.init_year])
+    clone.add_set("type_year", "baseyear_macro")
+    clone.add_set("cat_year", ["baseyear_macro", c.base_year])
+
+    # add nodal set structure
+    clone.add_set("type_node", "economy")
+    for n in c.nodes:
+        clone.add_set("cat_node", ["economy", n])
+
+    # add sectoral set structure
+    # TODO: we shouldn't have to have a for loop here
+    for s in c.sectors:
+        clone.add_set('sector', s)
+        clone.add_set("mapping_macro_sector", [s, s, "useful"])
+
+    # add parameters
+    for name, values in MACRO_INIT['pars'].items():
+        try:
+            key = values.get('data_key', name)
+            data = c.data[key].reset_index()
+            data['unit'] = values['unit']
+            # some data may have information prior to the MACRO initialization
+            # year which we need to remove in order to add it to the scenario
+            if 'year' in data:
+                data = data[data['year'] >= c.init_year]
```

Comment:
@gidden, I needed to add a check for `sector` here to prevent an error when using `add_par` for _demand_MESSAGE_:
```
            if 'sector' in data:
                data = data[data['sector'].isin(clone.set('sector'))]
```

Reply:
Added a check when reading in model data (as discussed in person)